cc/mcts_node.cc

```diff
@@ -0,0 +1,254 @@
+    if (it == node->children.end()) {
+      oss << "GAME END";
+      break;
+    }
+    node = it->second.get();
+    oss << node->move.ToKgs() << " (" << static_cast<int>(node->N())
+        << ") ==> ";
+  }
+  oss << std::setprecision(5) << "Q: " << node->Q();
+  return oss.str();
+}
+
+void MctsNode::InjectNoise(const std::array<float, kNumMoves>& noise) {
+  for (int i = 0; i < kNumMoves; ++i) {
+    edges[i].P = 0.75f * edges[i].P + 0.25f * noise[i];
+  }
+}
+
+MctsNode* MctsNode::SelectLeaf() {
+  auto* node = this;
+  for (;;) {
+    ++node->stats->N;
+
+    // If a node has never been evaluated, we have no basis to select a child.
+    if (node->state == State::kCollapsed) {
+      node->state = State::kSelected;
+      return node;
+    }
+    // If the node has already been selected for the next inference batch, we
+    // shouldn't select it again.
+    if (node->state == State::kSelected) {
+      // Revert the visits we just made.
+      --node->stats->N;
+      while (node != this) {
+        node = node->parent;
+        --node->stats->N;
+      }
+      return nullptr;
+    }
+    // HACK: if last move was a pass, always investigate double-pass first
+    // to avoid situations where we auto-lose by passing too early.
+    if (node->position.previous_move() == Coord::kPass &&
+        node->child_N(Coord::kPass) == 0) {
+      node = node->MaybeAddChild(Coord::kPass);
+      continue;
+    }
+
+    auto child_action_score = node->CalculateChildActionScore();
+    Coord best_move = ArgMax(child_action_score);
+    node = node->MaybeAddChild(best_move);
+  }
+}
+
+void MctsNode::IncorporateResults(absl::Span<const float> move_probabilities,
+                                  float value, MctsNode* up_to) {
+  assert(move_probabilities.size() == kNumMoves);
+  // A finished game should not be going through this code path, it should
+  // directly call BackupValue on the result of the game.
+  assert(!position.is_game_over());
+  assert(state == State::kSelected);
+  state = State::kExpanded;
+  for (int i = 0; i < kNumMoves; ++i) {
+    edges[i].original_P = edges[i].P = move_probabilities[i];
+    // Initialize child Q as current node's value, to prevent dynamics where
+    // if B is winning, then B will only ever explore 1 move, because the Q
+    // estimation will be so much larger than the 0 of the other moves.
+    //
+    // Conversely, if W is winning, then B will explore all 362 moves before
+    // continuing to explore the most favorable move. This is a waste of
+    // search.
+    //
+    // The value seeded here acts as a prior, and gets averaged into Q
+    // calculations.
+    edges[i].W = value;
+  }
+  BackupValue(value, up_to);
+}
+
+void MctsNode::IncorporateEndGameResult(float value, MctsNode* up_to) {
+  assert(position.is_game_over() || position.n() == kMaxSearchDepth);
+  assert(state == State::kSelected);
+  // We can't expand an end game result, so collapse the node again.
+  state = State::kCollapsed;
```

Comment:
oh, this is a clever solution to allow terminal nodes to be revisited repeatedly if necessary.

Reply:
A little clever, but I had two bugs in the first version of the code. Fortunately, the awesome Python unit tests caught them when I ported them to C++ :)